### ISSUE TITLE
Training bug when using `--sync_bn`

### ISSUE THREAD

**Issue by twotwoiscute** [ultralytics/yolov5#3754]
## ❔Question
Hi I have a problem with using `convert_sync_batchnorm` ,When I was trying to use DDP everything works fine ,but when I turn on the sync_bn mode ,the training process start and get stuck right away…

```
## Additional context
Here’s some info :
torch                         1.8.1                   
torchaudio                0.7.0                    
torchvision               0.9.1              

#How I run the script : 
python -m torch.distributed.launch  \
--nproc_per_node 4  \ 
--master_addr $master_addr  \ 
--master_port $port  train.py \ 
--train.py
--batch 256  --weights yolov5s.pt --device 0,1,2,3  \
--sync_bn

#How I init :
#note : opt.local_rank is -1 here 
    if opt.local_rank != -1:
        assert torch.cuda.device_count() > opt.local_rank
        torch.cuda.set_device(opt.local_rank)
        device = torch.device('cuda', opt.local_rank)
        dist.init_process_group(backend='nccl', init_method='env://')
```

 Thanks

**Comment by glenn-jocher** [comment]
@twotwoiscute your code is out of date, the opt class no longer carries DDP variables. To update your code:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by twotwoiscute** [comment]
@glenn-jocher Thanks for the suggestion 
However , I clone the latest version , However it still get stucked  at forward pass ,Here's what I did  : 
```
with amp.autocast(enabled=cuda):
    try:
        print("pred")
        pred = model(imgs)  # forward
        print("done pred")
        loss, loss_items = compute_loss(pred, targets.to(device))  # loss scaled by batch_size
        if RANK != -1:
            loss *= WORLD_SIZE  # gradient averaged between devices in DDP mode
        if opt.quad:
            loss *= 4.
    except Exception as error :
             print(error)
```
I found that if I use 4 gpus and the first  3 gpus would  have message `print("pred") `on the screen and the last gpu never print the message `"pred"` ,futhermore the first 3 gpus never enter to the code  `print("done pred") which means the forward pass never complete in any of gpus.

And this happen after first iteration completed.By the way,sync_bn would only fail if I use multiple gpus, with single gpu everything works perfectly.

**Comment by glenn-jocher** [comment]
@twotwoiscute ok understood, I will try to reproduce this today.

**Comment by glenn-jocher** [comment]
@twotwoiscute yes I can reproduce this. I'm not sure what the cause is. In looking at the documentation perhaps we are missing a `process_group` input to the convert function.
https://pytorch.org/docs/stable/generated/torch.nn.SyncBatchNorm.html

The convert function is here:
https://github.com/ultralytics/yolov5/blob/374957317a5469742b24291caa52dedfd9d31c99/train.py#L217-L221

**Comment by twotwoiscute** [comment]
I follow the  script you mentioned somehing like   :    
```
r1 = [ 0,1,2,3 ]
process_groups = [torch.distributed.new_group(pids) for pids in [r1] ] 
process_group = process_groups[0]

```
However ,it still does not work ...

**Comment by glenn-jocher** [comment]
@twotwoiscute I'm not sure if this is due to torch 1.9.0 or our own code, but I don't see anything wrong with the implementation. 

I would actually not use --sync if I were you though. It mainly helps in early training but we found it to have little to no effect on final mAP (provided you train long enough, i.e. 300 COCO epochs).

**Comment by twotwoiscute** [comment]
In my case , I have 64 batch size per gpu , I think it's enough for calculating running mean & var correctly . However,for completeness,maybe to let pytorch team to have a look at this issue ?

**Comment by glenn-jocher** [comment]
@twotwoiscute yes, that's a good idea. I would raise a bug report on the pytorch repo linking to this issue.

**Comment by glenn-jocher** [comment]
@twotwoiscute --sync known issue PR in #4032 to alert future users to the existing problem

**Comment by twotwoiscute** [comment]
> @twotwoiscute --sync known issue PR in #4032 to alert future users to the existing problem

Have you get any response from pytorch team? [I posted the similiar question in PyTorch Forums](https://discuss.pytorch.org/t/does-sync-batch-normalization-in-pytorch-not-work/129904) However, does not get any response..

**Comment by glenn-jocher** [comment]
@twotwoiscute we still have no resolution on this issue. If you find anything or hear from the PyTorch team please update here, thank you!

**Comment by twotwoiscute** [comment]
> @twotwoiscute we still have no resolution on this issue. If you find anything or hear from the PyTorch team please update here, thank you!
will do so :)

**Comment by twotwoiscute** [comment]
> @twotwoiscute we still have no resolution on this issue. If you find anything or hear from the PyTorch team please update here, thank you!

Will do so .

**Comment by twotwoiscute** [comment]
> @twotwoiscute we still have no resolution on this issue. If you find anything or hear from the PyTorch team please update here, thank you!

I think this issue  would be solved in current version since the way I solve this issue by commenting out  : 

`Thread(target=plot_images, args=(imgs, targets, paths, f), daemon=True).start()`

And `sync-bn` works perfectly fine.

**Comment by glenn-jocher** [comment]
@twotwoiscute do you mean you comment out L81 in loggers/__init__.py?
https://github.com/ultralytics/yolov5/blob/b894e69dfc341fcbfe4a307a15d6af90d90367df/utils/loggers/__init__.py#L79-L81

**Comment by glenn-jocher** [comment]
@twotwoiscute what about the daemon Thread plots in val.py here?
https://github.com/ultralytics/yolov5/blob/b894e69dfc341fcbfe4a307a15d6af90d90367df/val.py#L221-L227

**Comment by twotwoiscute** [comment]
@glenn-jocher  Ops! I think i made a wrong statement..  actually I comment out this part of code :
```
if tb_writer:
    tb_writer.add_graph(torch.jit.trace(de_parallel(model), imgs, strict=False), [])  # model graph
    # tb_writer.add_image(f, result, dataformats='HWC', global_step=epoch) 
```
but keep the ploting part .

**Comment by glenn-jocher** [comment]
@twotwoiscute I don't think this line correlates with the issue at all. It doesn't actually exist anymore, it's been replaced with loggers/__init__.py L137 that only runs on the `on_train_end()` callback, which will not have any effect on starting training with or without `--sync`:

https://github.com/ultralytics/yolov5/blob/b894e69dfc341fcbfe4a307a15d6af90d90367df/utils/loggers/__init__.py#L126-L138

**Comment by glenn-jocher** [comment]
@twotwoiscute wait are you saying the line you commented in your comment is uncommented, this line?
This line does run at train start, it's been moved to loggers/__init__.py L78. I can try commenting it, thanks for the tip!
https://github.com/ultralytics/yolov5/blob/b894e69dfc341fcbfe4a307a15d6af90d90367df/utils/loggers/__init__.py#L72-L78

**Comment by twotwoiscute** [comment]
@glenn-jocher okay,last weekend I tried to comment out  `plot_image `part and it works ,but today it gets stuck again...but when I comment out `self.tb.add_graph(torch.jit.trace(de_parallel(model), imgs[0:1], strict=False), [])` and works again ...

**Comment by glenn-jocher** [comment]
@twotwoiscute ok got it, thanks for the feedback! These are two very different things, add images is just adding images to TensorBoard, add_graph is adding an interactive YOLOv5 model visualizer (below), which is a much more complicated operation. I don't know what it has to do with --sync but if it's causing the hang we can simply not add graphs on --sync trainings as a workaround.

![image](https://user-images.githubusercontent.com/26833433/131326663-ff9ca48a-7071-4432-b0c7-b7e3d3f32759.png)

**Comment by glenn-jocher** [comment]
@twotwoiscute good news 😃! Your original issue may now be fixed ✅ in PR #4615. We discovered of the DPP `--sync-bn` issue was caused by TensorBoard add_graph() logging (used for visualizing the model interactively, example below). I don't know the exact cause and thus did not implement a fix, instead I implemented a workaround to avoid TensorBoard model visualization when `--sync-bn` is used. 

![131326663-ff9ca48a-7071-4432-b0c7-b7e3d3f32759](https://user-images.githubusercontent.com/26833433/131373309-49977b13-4bc4-46f1-a8c4-f725510bc420.png)

This means DDP training now works without issue with or without `--sync-bn`, but `--sync-bn` runs will not show a model visualization component in TensorBoard.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
